src/algorithms/mts.jl

```diff
@@ -0,0 +1,354 @@
+
+# MTS Algorithms
+# MTS: Multiple Trajectory Search for Large Scale Global Optimization, 
+# By [Lin-Yu Tseng and Chun Chen, 2008](https://sci2s.ugr.es/sites/default/files/files/TematicWebSites/EAMHCO/contributionsCEC08/tseng08mts.pdf)
+
+using Random: randperm
+
+@with_kw struct MTSOptions
+    method::Function = MTS
+    M = 100
+    maxiter=200
+    search_range_tol=1e-15
+    n_foreground = 80
+    n_local_search = 200
+    n_local_search_test = 3
+    n_local_search_best = 300
+    BONUS1 = 10
+    BONUS2 = 1 
+    a_min = 0.4
+    a_max = 0.5
+    b_min = 0.1
+    b_max = 0.3
+    c_min = 0
+    c_max = 1
+    function MTSOptions(args...; kwargs...)
+        @assert (length(args) > 0 && args[1] in _MTS_OPTIMIZATION_METHODS) || kwargs[:method] in _MTS_OPTIMIZATION_METHODS
+        new(args...; kwargs...)
+    end
+end
+
+@params mutable struct MTSWorkspace <: Workspace
+    model::VecModel
+    x0::AbstractVector
+    x::AbstractVector
+    options::MTSOptions
+    enable::AbstractVector
+    improve::AbstractVector
+    # Volitale variables
+    search_range::AbstractVector
+    optimal_x::AbstractVector
+    optimal_ind::Int
+    optimal_val::Real
+end
+function MTSWorkspace(model::VecModel, x0::AbstractVector, options::MTSOptions; kwargs...)
+    @unpack box_min, box_max = model
+    M = options.M
+    # Initialize improve and serch range
+    enable = [true for _ in 1:M]
+    improve = [true for _ in 1:M]
+    search_range = [(box_max-box_min) ./ 2 for _ in 1:M]
+    MTSWorkspace(model, x0, copy(x0), options, enable, improve, search_range, x0[1], -1, Inf)
+end
+
+if debugging[]
+    function Base.setproperty!(workspace::MTSWorkspace, name::Symbol, x)
+        if name in [:optimal_x, :optimal_ind, :optimal_val]
+            println("New ",  name, ": ", x)
+        end
+        setfield!(workspace, name, x)
+    end
+end
+
+@params struct MTSResult <: AbstractResult
+    minimum
+    minimizer
+end
+
+struct MTSAlg <: AbstractOptimizer end
+
+function reduce_search_range(search_range, k, n_dim, box_min, box_max, search_range_tol)
+    search_range[k] ./= 2
+    for i in 1:n_dim
+        if search_range[k][i] < search_range_tol
+            search_range[k][i] = (box_max[i] - box_min[i]) * 0.4
+        end
+    end
+end
+
+function _localsearch1(workspace, k)
+    @unpack model, options = workspace
+    @unpack x, improve, search_range = workspace
+    @unpack BONUS1, BONUS2, search_range_tol = options
+    @unpack box_min, box_max, n_dim = model
+    grade = 0
+    # search_range in paper is one-dimensional. Expand it to multidimensional. 
+    if improve[k] == false
+        reduce_search_range(search_range, k, n_dim, box_min, box_max, search_range_tol)
+    end
+    improve[k] = false
+    for i in 1:n_dim
+        # Original value
+        _xk = copy(x[k])
+        xk_val = evaluate!(model, _xk)
```

Comment:
if update_xki is true, this call is not needed because the previous value is valid, no?

Reply:
And if it's false, then we can cache the previous value as well so as not to re-evaluate.

Reply:
Yea I also realized this one when following the pseducode in paper, but I forget to optimize it. Thanks for pointing that out.